File: sdk_python/src/livepeer_frameworks/webhooks.py
"""Webhook verification and typed parsing. verify checks the Standard
Webhooks signature with the official standardwebhooks library; parse decodes
the body into the generated payload message of its event type. An event type
this SDK version does not know is returned as a raw event, so a new event type
never breaks an older SDK."""

from __future__ import annotations

import dataclasses
import json
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

import betterproto2
from standardwebhooks import Webhook, WebhookVerificationError as _StandardVerificationError

from ._generated.events import PUBLIC_EVENT_MESSAGES
from .errors import FrameWorksError, WebhookVerificationError

__all__ = ["WebhookEvent", "WebhookReceiver", "WebhookVerificationError", "parse_webhook_event", "PUBLIC_EVENT_MESSAGES"]

_SIGNATURE_HEADERS = ("webhook-id", "webhook-timestamp", "webhook-signature")


@dataclass(frozen=True)
class WebhookEvent:
    """One webhook event. id equals the webhook-id header; deduplicate on it.
    known is True when this SDK knows type: data is then the decoded payload
    message. For any other type data is None and raw_data holds the payload as
    received."""

    id: str
    type: str
    api_version: str
    created_at: str
    known: bool
    data: betterproto2.Message | None
    raw_data: Any


def _camel(name: str) -> str:
    head, *rest = name.split("_")
    return head + "".join(part.title() for part in rest)


def _sanitize(cls: type[betterproto2.Message], value: Any) -> Any:
    """Drops the fields and enum names a newer server may send that the
    generated message does not know, so decoding never fails on them."""
    if not isinstance(value, dict):
        return value
    message_cls: Any = cls
    fields = {f.name: f for f in dataclasses.fields(message_cls)}
    by_json = {_camel(name): name for name in fields}
    types: Mapping[str, Any] = message_cls._betterproto.cls_by_field
    out: dict[str, Any] = {}
    for key, item in value.items():
        name = by_json.get(key) or (key if key in fields else None)
        if name is None:
            continue
        meta = betterproto2.FieldMetadata.get(fields[name])
        field_type = types.get(name)
        if meta.proto_type == "enum" and isinstance(field_type, type):
            items = item if isinstance(item, list) else [item]
            kept = [v for v in items if _known_enum(field_type, v)]
            if meta.repeated:
                out[key] = kept
            elif kept:
                out[key] = item
            continue
        if meta.proto_type == "message" and meta.unwrap is None and isinstance(field_type, type):
            if isinstance(item, list):
                out[key] = [_sanitize(field_type, v) for v in item]
            else:
                out[key] = _sanitize(field_type, item)
            continue
        out[key] = item
    return out


def _known_enum(enum_cls: Any, value: Any) -> bool:
    if isinstance(value, int):
        return True
    # Webhooks carry the full proto names (ARTIFACT_KIND_CLIP); the generated
    # members drop the prefix (CLIP).
    if isinstance(value, str) and value in enum_cls.betterproto_renamed_proto_names_to_value():
        return True
    try:
        enum_cls.from_string(value)
    except (ValueError, KeyError):
        return False
    return True


def parse_webhook_event(payload: bytes | str) -> WebhookEvent:
    """Parses a webhook body into a typed event, without verifying it."""
    try:
        body = json.loads(payload)
    except ValueError as err:
        raise FrameWorksError("webhook body is not JSON") from err
    if not isinstance(body, dict):
        raise FrameWorksError("webhook body is not an event object")
    event_type = str(body.get("type") or "")
    raw = body.get("data")
    message_cls = PUBLIC_EVENT_MESSAGES.get(event_type)
    data = message_cls.from_dict(_sanitize(message_cls, raw or {})) if message_cls else None
    return WebhookEvent(
        id=str(body.get("id") or ""),
        type=event_type,
        api_version=str(body.get("api_version") or ""),
        created_at=str(body.get("created_at") or ""),
        known=message_cls is not None,
        data=data,
        raw_data=raw,
    )


class WebhookReceiver:
    """A receiver for one endpoint secret (whsec_...). Pass the raw request
    body: parsing and re-serializing JSON changes the bytes that were
    signed."""

    def __init__(self, secret: str) -> None:
        self._webhook = Webhook(secret)

    def verify(self, payload: bytes | str, headers: Mapping[str, str]) -> None:
        """Raises WebhookVerificationError unless the headers sign payload with
        this secret within five minutes."""
        signature_headers = {k.lower(): v for k, v in headers.items() if k.lower() in _SIGNATURE_HEADERS}
        try:
            self._webhook.verify(payload, signature_headers, json_parse=False)
        except _StandardVerificationError as err:
            raise WebhookVerificationError(str(err)) from err
        except ValueError as err:
            # A malformed signature header fails inside the library's parsing.
            raise WebhookVerificationError(f"invalid signature header: {err}") from err

    def receive(self, payload: bytes | str, headers: Mapping[str, str]) -> WebhookEvent:
        """verify, then parse_webhook_event."""
        self.verify(payload, headers)
        return parse_webhook_event(payload)

File: npm_api/src/webhooks.ts
/**
 * Webhook verification and typed parsing. verify checks the Standard
 * Webhooks signature with the official standardwebhooks library; parse
 * decodes the body into the generated payload type of its event type. An
 * event type this SDK version does not know is returned as a raw event, so a
 * new event type never breaks an older SDK.
 */
import {
  Webhook,
  WebhookVerificationError as StandardWebhookVerificationError,
} from "standardwebhooks";

import { FrameWorksError } from "./errors.js";
import {
  publicEventDecoders,
  type PublicEventDataMap,
  type PublicEventType,
} from "./generated/events.js";

export type { PublicEventDataMap, PublicEventType } from "./generated/events.js";
export * from "./generated/proto/events/public/v1/common.js";
export * from "./generated/proto/events/public/v1/access.js";
export * from "./generated/proto/events/public/v1/billing.js";
export * from "./generated/proto/events/public/v1/media.js";
export * from "./generated/proto/events/public/v1/streams.js";

/** The signature, timestamp, or headers of a webhook request did not verify. */
export class WebhookVerificationError extends FrameWorksError {
  constructor(message: string, cause?: unknown) {
    super(message, { cause });
    this.name = "WebhookVerificationError";
  }
}

interface EventEnvelope {
  /** Event ID, equal to the webhook-id header; deduplicate on it. */
  id: string;
  /** The payload version the endpoint pins, "v1". */
  apiVersion: string;
  /** When the change committed, RFC 3339. */
  createdAt: string;
}

/** An event of a type this SDK knows, with its payload decoded. */
export type KnownWebhookEvent = {
  [K in PublicEventType]: EventEnvelope & { known: true; type: K; data: PublicEventDataMap[K] };
}[PublicEventType];

/** An event of a type this SDK does not know; data is the JSON payload as received. */
export interface RawWebhookEvent extends EventEnvelope {
  known: false;
  type: string;
  data: unknown;
}

export type WebhookEvent = KnownWebhookEvent | RawWebhookEvent;

/** Request headers as fetch Headers or a Node IncomingHttpHeaders-like record. */
export type WebhookHeaders =
  | { get(name: string): string | null }
  | Readonly<Record<string, string | ReadonlyArray<string> | undefined>>;

const signatureHeaders = ["webhook-id", "webhook-timestamp", "webhook-signature"] as const;

function headerRecord(headers: WebhookHeaders): Record<string, string> {
  const out: Record<string, string> = {};
  if (typeof (headers as { get?: unknown }).get === "function") {
    const h = headers as { get(name: string): string | null };
    for (const name of signatureHeaders) {
      const v = h.get(name);
      if (v !== null) {
        out[name] = v;
      }
    }
    return out;
  }
  const record = headers as Readonly<Record<string, string | ReadonlyArray<string> | undefined>>;
  for (const [key, value] of Object.entries(record)) {
    const name = key.toLowerCase();
    if ((signatureHeaders as ReadonlyArray<string>).includes(name) && value !== undefined) {
      out[name] = typeof value === "string" ? value : value.join(" ");
    }
  }
  return out;
}

function payloadText(payload: string | Uint8Array): string {
  return typeof payload === "string" ? payload : new TextDecoder().decode(payload);
}

/** Parses a webhook body into a typed event, without verifying it. */
export function parseWebhookEvent(payload: string | Uint8Array): WebhookEvent {
  let body: unknown;
  try {
    body = JSON.parse(payloadText(payload));
  } catch (err) {
    throw new FrameWorksError("webhook body is not JSON", { cause: err });
  }
  if (body === null || typeof body !== "object") {
    throw new FrameWorksError("webhook body is not an event object");
  }
  const obj = body as Record<string, unknown>;
  const envelope = {
    id: String(obj.id ?? ""),
    apiVersion: String(obj.api_version ?? ""),
    createdAt: String(obj.created_at ?? ""),
  };
  const type = String(obj.type ?? "");
  const decoder = (
    publicEventDecoders as Record<string, { fromJSON(o: unknown): unknown } | undefined>
  )[type];
  if (!decoder) {
    return { ...envelope, known: false, type, data: obj.data };
  }
  return {
    ...envelope,
    known: true,
    type,
    data: decoder.fromJSON(obj.data ?? {}),
  } as KnownWebhookEvent;
}

export interface WebhookReceiver {
  /** Throws WebhookVerificationError unless the headers sign payload with this secret within five minutes. */
  verify(payload: string | Uint8Array, headers: WebhookHeaders): void;
  /** verify, then parseWebhookEvent. */
  receive(payload: string | Uint8Array, headers: WebhookHeaders): WebhookEvent;
}

/**
 * A receiver for one endpoint secret (whsec_...). Pass the raw request body:
 * parsing and re-serializing JSON changes the bytes that were signed.
 */
export function createWebhookReceiver(secret: string): WebhookReceiver {
  const wh = new Webhook(secret);
  const verify = (payload: string | Uint8Array, headers: WebhookHeaders) => {
    try {
      wh.verify(payloadText(payload), headerRecord(headers), { jsonParse: false });
    } catch (err) {
      if (err instanceof StandardWebhookVerificationError) {
        throw new WebhookVerificationError(err.message, err);
      }
      throw err;
    }
  };
  return {
    verify,
    receive(payload, headers) {
      verify(payload, headers);
      return parseWebhookEvent(payload);
    },
  };
}
